### PR TITLE
feat: create complete profile

### PR DESCRIPTION
**Pull request by ProggerOleg**
## Describe your changes
- Created entities for complete profile step (certificate, work_experience, caregiver_info )
-  Created main endpoints for complete-profile step
- Created enums for qualification page and availability page
- Created service for uploading caregivers video and saving it on S3 bucket

Screenshot from 2023-11-28 18-57-59

Screenshot from 2023-11-26 12-38-16

## Issue ticket code (and/or) and link
- [Link to JIRA ticket](#https://homecareaid.atlassian.net/jira/software/projects/AC/boards/1?selectedIssue=AC-105)

### **General**
- [x] Assigned myself to the PR
- [x] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [x] Updated the documentation
- [x] Performed a self-review of my code
- [x] Types for input and output parameters
- [x] Don't have "any" on my code
- [x] Used the try/catch pattern for error handling
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside the ternary operator
- [x] Don't have commented code
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [x] Used camelCase for variables and functions
- [x] Date and time formats are on the constants
- [x] Functions are public only if it's used outside the class
- [x] No hardcoded values
- [ ] covered by tests
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

### Backend
- [x] Swagger documentation updated
- [x] Database requests are optimized and not redundant
- [ ] Unit tests written
- [x] use ConfigService instead of process.env
- [x] use transactions if there is a call chain that mutates data in different tables
- [ ] use @index decorator for frequently requested data

### CODE REVIEW

**Comment by ProggerOleg** [comment]
Lyudmila, please could you look at my data structure, I’m not sure that this is the best way to build dependencies, should work experience and certificates be saved into separate tables or it would be better to turn them into json?

**Comment by lyudad** [comment]
Screenshot 2023-11-24 at 20 34 03

**Comment by ProggerOleg** [comment]
Fixed keys, but i believe we need another table for caregivers info, because other way there will be lots of empty fields for users who are seekers and don't need fields with qualification, etc

File: src/modules/profile/profile.controller.ts [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+import { CreateCertificatesDto } from './dto/create-certificate.dto';
+import { CreateWorkExperienceDto } from './dto/create-work-experience.dto';
+import { WorkExperienceDto } from './dto/work-experience.dto';
+
+@ApiTags('Complete profile')
+@Controller('profile')
+export class ProfileController {
+  constructor(private readonly profileService: ProfileService) {}
+
+  @Post('/uploadFile/:userId')
```
Comment (Simpsonkk) [inline src/modules/profile/profile.controller.ts]: Could you please use an enum for all API paths? (as we did in other controllers)

File: src/modules/profile/profile.controller.ts [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+        ],
+      }),
+    )
+    file: Express.Multer.File,
+    @Param('userId') userId: string,
+  ): Promise<void> {
+    await this.profileService.upload(file.originalname, file.buffer, userId);
+  }
+
+  @Get('/:userId')
```
Comment (Simpsonkk) [inline src/modules/profile/profile.controller.ts]: same here

File: src/modules/profile/profile.controller.ts [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    status: HttpStatus.INTERNAL_SERVER_ERROR,
+    description: ErrorMessage.InternalServerError,
+  })
+  async getProfileInformation(
+    @Param('userId') userId: string,
+  ): Promise<CaregiverInfo | undefined> {
+    return this.profileService.getProfileInformation(userId);
+  }
+
+  @Get('/work-experience/:userId')
```
Comment (Simpsonkk) [inline src/modules/profile/profile.controller.ts]: same here

File: src/modules/profile/profile.controller.ts [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    status: HttpStatus.INTERNAL_SERVER_ERROR,
+    description: ErrorMessage.InternalServerError,
+  })
+  async getWorkExperiences(
+    @Param('userId') userId: string,
+  ): Promise<WorkExperience[]> {
+    return this.profileService.getWorkExperiences(userId);
+  }
+
+  @Get('/certificates/:userId')
```
Comment (Simpsonkk) [inline src/modules/profile/profile.controller.ts]: same here

File: src/modules/profile/profile.controller.ts [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    status: HttpStatus.INTERNAL_SERVER_ERROR,
+    description: ErrorMessage.InternalServerError,
+  })
+  async getUserCertificates(
+    @Param('userId') userId: string,
+  ): Promise<Certificate[]> {
+    return this.profileService.getUserCertificates(userId);
+  }
+
+  @Post(':userId')
```
Comment (Simpsonkk) [inline src/modules/profile/profile.controller.ts]: same here

File: src/modules/profile/profile.controller.ts [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+  })
+  @ApiResponse({
+    status: HttpStatus.INTERNAL_SERVER_ERROR,
+    description: ErrorMessage.InternalServerError,
+  })
+  async createProfile(@Param('userId') userId: string): Promise<void> {
+    await this.profileService.createProfile(userId);
+  }
+
+  @Patch('/:userId')
```
Comment (Simpsonkk) [inline src/modules/profile/profile.controller.ts]: same here

File: src/modules/profile/profile.controller.ts [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    description: ErrorMessage.InternalServerError,
+  })
+  async updateProfile(
+    @Param('userId') userId: string,
+    @Body() updateProfileDto: UpdateProfileDto,
+  ): Promise<void> {
+    await this.profileService.updateProfile(userId, updateProfileDto);
+  }
+
+  @Post('/certificates/:userId')
```
Comment (Simpsonkk) [inline src/modules/profile/profile.controller.ts]: same here

File: src/modules/profile/profile.controller.ts [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+    description: ErrorMessage.InternalServerError,
+  })
+  addCertificate(
+    @Param('userId') userId: string,
+    @Body() createCertificateDto: CreateCertificatesDto,
+  ): Promise<Certificate[]> {
+    return this.profileService.createCertificate(userId, createCertificateDto);
+  }
+
+  @Post('/work-experience/:userId')
```
Comment (Simpsonkk) [inline src/modules/profile/profile.controller.ts]: same here

File: src/modules/profile/profile.service.ts [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+    if (hourlyRate) {
+      caregiverInfo.hourlyRate = hourlyRate;
+    }
+
+    if (description) {
+      caregiverInfo.description = description;
+    }
+
+    await this.profileRepository.save(caregiverInfo);
```
Comment (Simpsonkk) [inline src/modules/profile/profile.service.ts]: Could we do something like this? I mean just put data in method .update(updateData)
example:
async update(email: string, userDto: Partial<User>): Promise<void> {
    try {
      await this.userRepository
        .createQueryBuilder()
        .update(User)
        .set(userDto)
        .where('user.email = :email', {
          email,
        })
        .execute();
    } catch (error) {
      throw new HttpException(error.message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }
}

File: src/modules/profile/profile.controller.ts [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+  }
+
+  @Post(ProfileApiPath.WorkExperience)
+  @ApiOperation({ summary: 'Add work experience(s)' })
+  @ApiResponse({
+    status: HttpStatus.CREATED,
+    description: 'Work experience added successfully',
+    isArray: true,
+    schema: {
+      example: [
```
Comment (lyudad) [inline src/modules/profile/profile.controller.ts]: вынеси пример в константу в  какой то файл с примером, в этой же папке, он отвлекает от важного, хотя и важный

File: src/modules/profile/profile.controller.ts [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+import { ProfileService } from 'src/modules/profile/profile.service';
+
+import { MAX_FILE_SIZE } from './constants/complete-profile.constants';
+import { UpdateProfileDto } from './dto/additional-info.dto';
+import { CreateCertificatesDto } from './dto/create-certificate.dto';
+import { CreateWorkExperienceDto } from './dto/create-work-experience.dto';
+import { WorkExperienceDto } from './dto/work-experience.dto';
+import { ProfileApiPath } from './enum/profile.enum';
+
+@ApiTags('Complete profile')
```
Comment (lyudad) [inline src/modules/profile/profile.controller.ts:45]: не могу понять то ли я не вижу то ли не сделано. а где защита по токену контролерра, он же не публичный на сколько я понимаю

File: src/modules/profile/profile.controller.ts (hunk 10 above)
Comment (ProggerOleg) [reply src/modules/profile/profile.controller.ts]: Fixed

File: src/modules/profile/profile.controller.ts (hunk 11 above)
Comment (ProggerOleg) [reply src/modules/profile/profile.controller.ts:45]: Fixed, created auth guard in auth folder

**Review by lyudad — APPROVED** [review]

**Pull request merged**